src/graphnet/models/detector/detector.py

```diff
@@ -1,115 +1,53 @@
+    def feature_map(self) -> Dict[str, Callable]:
+        """List of features used/assumed by inheriting `Detector` objects."""
 
     @final
-    def forward(self, data: Data) -> Data:
+    def forward(  # type: ignore
+        self, node_features: torch.tensor, node_feature_names: List[str]
+    ) -> Data:
         """Pre-process graph `Data` features and build graph adjacency."""
-        # Check(s)
-        assert data.x.size()[1] == self.nb_inputs, (
-            "Got graph data with incompatible size, ",
-            f"{data.x.size()} vs. {self.nb_inputs} expected",
-        )
-
-        # Graph-bulding
-        # @NOTE: `.clone` is necessary to avoid modifying original tensor in-place.
-        data = self._graph_builder(data).clone()
-
-        if self._scalers:
-            # # Scaling individual features
-            # x_numpy = data.x.detach().cpu().numpy()
-            # for key, scaler in self._scalers.items():
-            #     ix = self.features.index(key)
-            #     data.x[:,ix] = torch.tensor(scaler.transform(x_numpy[:,ix])).type_as(data.x)
-
-            # Scaling groups of features | @TEMP, probably
-            x_numpy = data.x.detach().cpu().numpy()
+        return self._standardize(node_features, node_feature_names)
 
-            data.x[:, :3] = torch.tensor(
-                self._scalers["xyz"].transform(x_numpy[:, :3])  # type: ignore[call-overload]
-            ).type_as(data.x)
-
-            data.x[:, 3:] = torch.tensor(
-                self._scalers["features"].transform(x_numpy[:, 3:])  # type: ignore[call-overload]
-            ).type_as(data.x)
-
-        else:
-            # Implementation-specific forward pass (e.g. preprocessing)
-            data = self._forward(data)
-
-        return data
-
-    @abstractmethod
-    def _forward(self, data: Data) -> Data:
-        """Syntax like `.forward`, for implentation in inheriting classes."""
-
-    @property
-    def nb_inputs(self) -> int:
-        """Return number of input features."""
-        return len(self.features)
-
-    @property
-    def nb_outputs(self) -> int:
-        """Return number of output features.
-
-        This the default, but may be overridden by specific inheriting classes.
-        """
-        return self.nb_inputs
-
-    def _validate_features(self, data: Data) -> None:
-        if isinstance(data, Batch):
-            # `data.features` is "transposed" and each list element contains only duplicate entries.
-
-            if (
-                len(data.features[0]) == data.num_graphs
-                and len(set(data.features[0])) == 1
-            ):
-                data_features = [features[0] for features in data.features]
+    @final
+    def _standardize(
+        self, node_features: torch.tensor, node_feature_names: List[str]
+    ) -> Data:
+        for idx, feature in enumerate(node_feature_names):
+            try:
+                node_features[:, idx] = self.feature_map()[feature](  # type: ignore
```

Comment:
I'm not sure if you're supposed to call self.feature_map, since it is classified as a property?